README.md
# 📜 Curtains
## [pub.dev Listing](https://pub.dev/packages/curtains) | [API Doc](https://pub.dev/documentation/curtains/latest) | [GitHub](https://github.com/Zabadam/curtains)
#### API References: [`Curtains`](https://pub.dev/documentation/curtains/latest/curtains/Curtains-class.html) | [`Sensitivity`](https://pub.dev/documentation/curtains/latest/curtains/Sensitivity-class.html) | 👥 [`package:shadows`](https://pub.dev/packages/shadows): 🕴 [`Elevation`](https://pub.dev/documentation/shadows/latest/shadows/Elevation/asBoxDecoration.html)

<br />

[![](https://raw.githubusercontent.com/Zabadam/curtains/main/doc/Curtains.png)](https://pub.dev/documentation/curtains/latest/curtains/curtains-library.html 'pub.dev Documentation')

Wrap a scrollable with scrim 📜 `Curtains` while not at the start or end, alluding to unrevealed content.

<br />

| [![](https://raw.githubusercontent.com/Zabadam/curtains/main/doc/elevated.gif 'Link to the first demo in the Example app')](https://github.com/Zabadam/curtains/blob/main/example/lib/main.dart#L222 'First demo in the Example source') | [![](https://raw.githubusercontent.com/Zabadam/curtains/main/doc/code.png 'Not the exact code running in the gif next door, but it is this easy! Click for full size.')](https://raw.githubusercontent.com/Zabadam/curtains/main/doc/code.png) |
| :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |

<br />

> There has been a breaking change from **[0.9.4+1]** to **[0.9.5]**,
> but it is quite easily fixed. \
> Package still in pre-release.
> 
> A few other small changes have been made that may require minor cleanup \
> if moving to this version from an older one.
> 
> | <h2 style='color:green'>**[0.10.0] (now)**</h2> |   ⬅️   | <h2 style='color:orange'>**[0.9.4+1] (before)**</h2> | \|   | <h2 style='color:green'>**[0.10.0] (now)**</h2> | ⬅️    | <h2 style='color:orange'>**[0.9.4+1] (before)**</h2> |
> | :---------------------------------------------- | :---: | :------------------------------------------------ | :--- | :---------------------------------------------- | :--- | :------------------------------------------------ |
> | `Curtains`                                      |       | `Curtains.regal`                                  | \|   | `Sensitivity sensitivity`                       |      | `List<double> sensitivity`                        |
> | `Curtains.instant`                              |       | `Curtains.fancy`                                  | \|   | `TextDirection textDirection`                   |      | `TextDirection directionality`                    |
> | `Curtains.elevated`                             |       | `Curtains`                                        | \|   | `bool endInitVisible` <h6>⭐</h6>              |      | `bool endCurtainInitVisible`                      |
> 
> #### ⭐ Also now available is `bool startInitVisible`.
> <br />
> 
> Furthermore either `startCurtain` or `endCurtain` [now accepts a more
> abstract `Decoration?`](#any-decoration) \
> as opposed to a restriction to `BoxDecoration?`.

---

## 📚 Table of Contents
- 📜 [Curtains](#-curtains)
- 🧫 [Examples](#-examples-table-of-contents-scroll-up-to--table-of-contents)
  - [Elevated 📜 `Curtains`](#elevated--curtains-table-of-contents-scroll-up-to--table-of-contents)
  - [Instant 📜 `Curtains`](#instant--curtains-table-of-contents-scroll-up-to--table-of-contents)
  - [Animated 📜 `Curtains`](#animated--curtains-table-of-contents-scroll-up-to--table-of-contents)
- 📖 [Reference](#-reference-table-of-contents-scroll-up-to--table-of-contents)
  - 🕴 [Elevation](#-elevation-table-of-contents-scroll-up-to--table-of-contents)
  - [Any `Decoration`](#any-decoration-table-of-contents-scroll-up-to--table-of-contents)
- 🐞 [Bugs](#-bugs-table-of-contents-scroll-up-to--table-of-contents)
- 🐸 [More by Zaba](#-zabaapp--simple-packages-simple-nameshttpspubdevpublisherszabaapppackages-other-flutter-packages-published-by-zabaapp)

---

# 🧫 [Examples](#-table-of-contents 'Scroll up to 📚 Table of Contents')
See demonstrations of this Flutter package:
- [📜 `Curtains` Demo source: `/example/lib/main.dart`](https://github.com/Zabadam/curtains/tree/main/example/lib/main.dart '📜 Curtains Demo source: `/example/lib/main.dart`')
- [📜 `Curtains` Demo APK](https://github.com/Zabadam/curtains/blob/main/example/build/app/outputs/apk/release/app-release.apk '📜 Curtains Demo APK')

## [Elevated 📜 `Curtains`](#-table-of-contents 'Scroll up to 📚 Table of Contents')
In lieu of `Decoration`s, provide a simple number-based `elevation`.
Optionally provide a `color` or specify `duration`.

```dart
final curtains = Curtains.elevated(
  child: ListView(
    children: List.generate(
      25,
      (i) => ListTile(title: Text('ListTile #: ${i+1}')),
    ),
  ),
  elevation: 24.0 // optional, defaults to `9.0`
  // Optional color; will maintain Material elevation opacities.✝
  color: Colors.red
);
```
> ## [It's *that* easy!](https://github.com/Zabadam/curtains/blob/main/example/lib/main.dart#L222 'First demo in the Example source. Since recording, Shadows.elevated are now animated by default!')
>
> #### ✝ [👥 `Shadows`](https://pub.dev/packages/shadows) - [See: `Elevation.asBoxShadows`](https://pub.dev/documentation/shadows/latest/shadows/Elevation/asBoxShadows.html)

&nbsp;

## [Instant 📜 `Curtains`](#-table-of-contents 'Scroll up to 📚 Table of Contents')
Special `Curtains` that offer no animation support. \
No `AnimationController`s were hurt in the making of this widget.

```dart
final curtains = Curtains.instant(
  /// 🕴 Default constructor 📜 [Curtains] employs `Elevation.asBoxDecoration` to
  /// render its decorations; but feel free to use these static methods, too,
  /// if you opt for 👥 [package:shadows].
  startCurtain: const BoxDecoration(. . .) // Elevation.asBoxDecoration(12.0), // 🕴
  endCurtain: BoxDecoration(boxShadow: Elevation.asBoxShadows(12.0)), // 🕴
  child: ListView(
    children: List.generate(
      25,
      (i) => ListTile(title: Text('ListTile #${i+1}')),
    ),
  ),
);
```
> ## [It's *similarly* easy!](https://github.com/Zabadam/curtains/blob/main/example/lib/main.dart#L242 'Second demo in the Example source')

&nbsp;

## [Animated 📜 `Curtains`](#-table-of-contents 'Scroll up to 📚 Table of Contents')
Each curtain may be customized independently. The `spread` parameter provides
"girth," either width or height depending on axis, to allow the application
of `Decoration` images, gradients, and fills. 

Without `spread`, the only truly valid form of scrim curtain is the
`List<BoxShadow>` from `BoxDecoration`, `ShapeDecoration`, etc.

| [![](https://raw.githubusercontent.com/Zabadam/curtains/main/doc/sensitivity.gif 'Custom animated BoxDecoration scrim Curtains and initialized `sensitivity`')](https://github.com/Zabadam/curtains/blob/main/example/lib/main.dart#L390 'Penultimate demo in the Example source') | [This instance of 📜 `Curtains`](https://github.com/Zabadam/curtains/blob/main/example/lib/main.dart#L363 'Final demo in the Example source') has </br>custom, animated BoxDecorations <br />as well as [`sensitivity` at the start <br />and end](https://pub.dev/documentation/curtains/latest/curtains/Sensitivity-class.html). |
| :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: | :-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

```dart
BoxDecoration buildCurtain({bool isStart = true}) => BoxDecoration(
  gradient: LinearGradient(
    colors: [Colors.green[400]!, Colors.green[400]!.withOpacity(0)],
    begin: (isStart) ? Alignment.centerLeft : Alignment.centerRight,
    end: (isStart) ? Alignment.centerRight : Alignment.centerLeft,
  ),
  boxShadow: const [
    BoxShadow(color: Color(0x22FF0000), spreadRadius: 0.0, blurRadius: 5.0),
    BoxShadow(color: Color(0x66FF0000), spreadRadius: 10.0, blurRadius: 30.0),
    BoxShadow(color: Color(0x22FF0000), spreadRadius: 60.0, blurRadius: 150.0),
  ],
);

final curtains = Curtains(
  startCurtain: buildCurtain(),
  endCurtain: buildCurtain(isStart: false),
  // Provide `spread` to [Curtains] for decoration support beyond [BoxShadow]s.

CHANGELOG.md
# 📜 Curtains
##### **WORK IN PROGRESS**
## **[0.10.0] - 31 MAY 21**
- Code and documentation cleanup.
- Constructor and parameter alterations:
  
> | <h3 style='color:green'>**[0.10.0] (now)**</h3> |   ⬅️   | <h3 style='color:red'>**[0.9.4+1] (before)**</h3> | \|   | <h3 style='color:green'>**[0.10.0] (now)**</h3> | ⬅️    | <h3 style='color:red'>**[0.9.4+1] (before)**</h3> |
> | :---------------------------------------------- | :---: | :------------------------------------------------ | :--- | :---------------------------------------------- | :--- | :------------------------------------------------ |
> | `Curtains`                                      |       | `Curtains.regal`                                  | \|   | `Sensitivity sensitivity`                       |      | `List<double> sensitivity`                        |
> | `Curtains.instant`                              |       | `Curtains.fancy`                                  | \|   | `TextDirection textDirection`                   |      | `TextDirection directionality`                    |
> | `Curtains.elevated`                             |       | `Curtains`                                        | \|   | `bool endInitVisible` ⭐                         |      | `bool endCurtainInitVisible`                      |
> 
> ⭐ Also now available is `bool startInitVisible`.
> <br /><br />
> Furthermore either `startCurtain` or `endCurtain` [now accepts a more
> abstract `Decoration?`](https://pub.dev/packages/curtains/#any-decoration-table-of-contents-scroll-up-to--table-of-contents) \
> as opposed to the restriction to `BoxDecoration?`.
<br />

## **[0.9.4+1] - 04 MAY 21**
- Removed erroneous `print()`

## **[0.9.4] - 30 APR 21**
- Added 🆕 parameters: `endCurtainInitVisible`, which should be self explanatory; and `alignment`, similarly so.

## **[0.9.3+1] - 25 APR 21**
- Corrected `export 'package:flutter/widgets.dart';`.

## **[0.9.3] - 24 APR 21**
- Divided package into two separate entities:
  - First is [`📜 Curtains`](https://pub.dev/packages/curtains), this package, for super simple scroll scrims.
  - 🆕 Now available individually to handle `Elevation` paradigm is [`👥 package:shadows`](https://pub.dev/packages/shadows).
    - Check it out for some neat [`Box`](https://pub.dev/documentation/shadows/latest/shadows/Elevation/asBoxShadows.html)[`Shadow`](https://pub.dev/documentation/shadows/latest/shadows/BoxShadowUtils.html) [utilities](https://pub.dev/documentation/shadows/latest/shadows/BoxShadowsUtils.html) that mesh with 📜 `Curtains`.
- 🆕 Parameter: `color` for `Curtains`.
- `Curtains.elevation` parameter is now decoupled from the `key`s in `kElevationToShadow` as  [`Elevation.asBoxShadows`](https://pub.dev/documentation/shadows/latest/shadows/Elevation/asBoxShadows.html) will `lerp` if necessary.

## **[0.9.2] - 23 APR 21**
- 🐞 Bug fixed: `Curtains.regal` now properly respect initialized `duration` and `curve`.

## **[0.9.1] - 23 APR 21**
- 🆕 Parameter added: [`Curtains.clipBehavior`](https://pub.dev/documentation/curtains/latest/curtains/Curtains/clipBehavior.html 'pub.dev Documentation: Curtains.clipBehavior')
  - Default (and previous) behavior: `Clip.hardEdge` for easy `BoxShadow` support (otherwise they bleed over)
  - Initialize `Clip.none` to alter Widget tree depth & remove [`ClipRect`](https://api.flutter.dev/flutter/widgets/ClipRect-class.html 'Flutter API ref doc: ClipRect') from 📜 Curtains

## **[0.9.0]/[0.9.0+1] - 23 APR 21**
- Initial release. Supports sound null-safety.
- One known 🐞 **bug**:
  - [Even with `Curtains.spread` set non-negligibly](https://github.com/Zabadam/curtains/blob/main/example/lib/main.dart#L319 '/example/lib/main.dart#L319'), [`BoxDecoration.backgroundBlendMode` does not work correctly](https://github.com/Zabadam/curtains/blob/main/example/lib/main.dart#L379 '/example/lib/main.dart#L379').


example/lib/source_code.dart
/// Demo source code in a bespoke `CodeViewer` itself employing [Curtains].
library curtains_demo;

import 'package:flutter/material.dart';

import 'package:curtains/curtains.dart';

/// {@macro code_viewer}
class CodeViewer extends StatelessWidget {
  /// {@template code_viewer}
  /// An [InteractiveViewer] meant to display source code.
  ///
  /// Beginning and end of this scrollable use [Curtains] as shadow scrims.
  /// {@endtemplate}
  const CodeViewer(this.data, {Key? key}) : super(key: key);

  /// A raw `String` of source code to present within this `CodeViewer`.
  final String data;

  @override
  Widget build(BuildContext context) {
    return InteractiveViewer(
      maxScale: 4.0,
      child: DecoratedBox(
        decoration: const BoxDecoration(color: Color(0xFF3C2121)),
        child: Curtains(
          startCurtain: const BoxDecoration(
            boxShadow: [
              BoxShadow(
                color: Color(0x44FF4444),
                spreadRadius: 20,
                blurRadius: 40,
              )
            ],
          ),
          endCurtain: const BoxDecoration(
            boxShadow: [
              BoxShadow(
                color: Color(0xAA990000),
                spreadRadius: 20,
                blurRadius: 40,
              )
            ],
          ),
          sensitivity: const Sensitivity(15, 40),
          child: SingleChildScrollView(
            physics: const BouncingScrollPhysics(),
            padding: const EdgeInsets.only(
              left: 75,
              top: 15,
              right: 15,
              bottom: 15,
            ),
            child: SelectableText.rich(
              TextSpan(
                text: data,
                style: const TextStyle(
                    color: Colors.white,
                    fontSize: 10,
                    fontFamily: 'monospace',
                    fontWeight: FontWeight.bold,
                    letterSpacing: 0,
                    height: 1.5,
                    shadows: [Shadow(offset: Offset(0, 1))]),
              ),
            ),
          ),
        ),
      ),
    );
  }
}

/// {@macro source_code}
class SourceCode extends StatelessWidget {
  /// {@template source_code}
  /// Returns a `const` [CodeViewer] with a hard-coded raw `String` containing
  /// all source code from the demonstration proper.
  /// {@endtemplate}
  const SourceCode({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const CodeViewer(
      r'''
/// ## 📜 Curtains Demonstration
library curtains_demo;

import 'dart:math';
import 'package:flutter/services.dart';
import 'package:flutter/material.dart';

import 'package:curtains/curtains.dart';

import 'source_code.dart';

/// This demonstration.
late BuildContext curtainsDemo;

///     const DURATION = Duration(milliseconds: 600);
// ignore: constant_identifier_names
const DURATION = Duration(milliseconds: 600);

///     const CURVE = Curves.fastOutSlowIn;
// ignore: constant_identifier_names
const CURVE = Curves.fastOutSlowIn;

/// Ordered List of demo names.
// ignore: constant_identifier_names
const TITLES = [
  '📜 Curtains Demo - Elevated',
  '📜 Instant Curtains: Axis.horizontal',
  '📜 Curtains: Animations',
  '📜 Curtains: BoxDecorations',
  '📜 Curtains: Curtain Spread & clipBehavior',
  '📜 Curtains: Sensitivity',
  '📜 Curtains: Any Decoration',
];

/// Starts at `TITLES[0]`, '📜 Curtains Demo', but changes as the
/// `>` IconButton over [CurrentDemo] is tapped.
var title = TITLES[0];

/// Hover in most IDEs for context tooltip
/// describing the significance of each Demo.
// ignore: constant_identifier_names
const DEMOS = [
  ElevatedCurtainsDemo(),
  InstantCurtainsDemo(),
  CurtainsDemoVertical1(),
  CurtainsDemoVertical2(),
  CurtainsDemoHorizontal1(),
  CurtainsDemoVertical3(),
  CurtainsDemoHorizontal2(),
];

/// Starts at `0`, [ElevatedCurtainsDemo], but changes as the
/// `>` IconButton over [CurrentDemo] is tapped.
var currentDemo = 0;

void main() => runApp(const CurtainsDemo());

/// {@macro curtains_demo}
class CurtainsDemo extends StatefulWidget {
  /// {@template curtains_demo}
  /// A [MaterialApp] frame for this entire demonstration.
  /// {@endtemplate}
  const CurtainsDemo({Key? key}) : super(key: key);

  @override
  _CurtainsDemoState createState() => _CurtainsDemoState();
}

class _CurtainsDemoState extends State<CurtainsDemo> {
  @override
  Widget build(BuildContext context) {
    curtainsDemo = context; // Allows easy rebuild from [CurrentDemo]
    return MaterialApp(
      title: title,
      theme: ThemeData(primarySwatch: Colors.red),
      home: SafeArea(
        child: Scaffold(
          appBar: AppBar(
            title: Text(title),
            automaticallyImplyLeading: false,
            toolbarHeight: (currentDemo == -1) ? 80 : null,
          ),
          body: const CurrentDemo(),
        ),
      ),
    );
  }
}

/// {@macro current_demo}
class CurrentDemo extends StatefulWidget {
  /// {@template current_demo}
  /// Setup the body of [CurtainsDemo] Scaffold
  /// according to the [Axis] of the [currentDemo].
  ///
  /// Provides an > `IconButton` to change [currentDemo].
  /// {@endtemplate}
  const CurrentDemo({Key? key}) : super(key: key);

  @override
  _CurrentDemoState createState() => _CurrentDemoState();
}

class _CurrentDemoState extends State<CurrentDemo> {
  @override
  Widget build(BuildContext context) {
    final isHorizontal =
        (currentDemo != 1 && currentDemo != 4 && currentDemo != 6);

    final children = <Widget>[
      isHorizontal ? headerVertical : headerHorizontal,
      DEMOS[(currentDemo == -1) ? 0 : currentDemo],
      isHorizontal ? footerVertical : footerHorizontal,
    ];

    return WillPopScope(
      onWillPop: () async {
        if (currentDemo == -1) {
          setState(() {
            currentDemo = 0;
            title = TITLES[0];
            (curtainsDemo as Element).markNeedsBuild();
          });
          return false;
        }
        var willPop = false;
        Scaffold.of(context).showBottomSheet<void>(
          (_) => GestureDetector(
            onTap: () => Navigator.pop(context),
            child: Container(
              height: 200,
              color: Colors.red[900],
              child: Center(
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                  children: <Widget>[
                    const Text(
                      'Exit 📜 Curtains Demo?',
                      style: TextStyle(
                        color: Colors.white,
                        fontSize: 40,
                      ),
                    ),
                    Row(
                      mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                      children: [
                        ElevatedButton(
                          onPressed: () {
                            willPop = true;
                            SystemChannels.platform.invokeMethod(
                              'SystemNavigator.pop',
                            );
                          },
                          child: const Text('\n  E X I T  \n'),
                        ),
                        ElevatedButton(
                          onPressed: () => Navigator.pop(context),
                          child: const Text('\n  S T A Y  \n'),
                        ),
                      ],
                    ),
                  ],
                ),
              ),
            ),
          ),
        );
        return willPop;
      },
      child: Stack(
        children: <Widget>[
          (currentDemo == -1)
              ? const SourceCode()
              : isHorizontal
                  ? Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      mainAxisAlignment: MainAxisAlignment.spaceBetween,
                      children: children,
                    )
                  : Row(children: children),
          IconButton(
            icon: const Icon(Icons.arrow_forward_ios),
            color: (currentDemo == -1) ? Colors.white : Colors.black,
            iconSize: 70,
            onPressed: () => setState(
              () {
                HapticFeedback.vibrate();
                currentDemo =
                    (currentDemo == DEMOS.length - 1) ? 0 : currentDemo + 1;
                title = TITLES[currentDemo];
                (curtainsDemo as Element).markNeedsBuild();
              },
            ),
          ),
          Positioned(
            top: 0,
            right: isHorizontal
                ? (currentDemo == -1)
                    ? -1920
                    : 16
                : -4,
            child: IconButton(
              icon: const Icon(Icons.code),
              color: isHorizontal ? Colors.red : Colors.blue,
              iconSize: isHorizontal ? 70 : 50,
              onPressed: () => setState(() {
                HapticFeedback.vibrate();
                currentDemo = -1;
                title = '📜 Curtains Demo: Source Code\n'
                    '🔎 Pinch to Zoom  👆 Tap and Hold to Select';
                (curtainsDemo as Element).markNeedsBuild();
              }),
            ),
          ),
        ],
      ),
    );
  }
}

/// 🕴 Consider [Material.elevation], but see [Elevation].
class ElevatedCurtainsDemo extends StatelessWidget {
  /// 🕴 Consider [Material.elevation], but see [Elevation].
  const ElevatedCurtainsDemo({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Expanded(
      child: Curtains.elevated(
        elevation: 24, // Consider [Material.elevation], but see [Elevation].🕴
        // endCurtainInitVisible: false,
        child: ListView(
          physics: const BouncingScrollPhysics(),
          children: generatedListVertical,
        ),
      ),
    );
  }
}

/// {@macro instant_curtains}
class InstantCurtainsDemo extends StatelessWidget {
  /// {@template instant_curtains}
  /// ↔ When instantiating, 📜 [Curtains] may be created on [Axis.horizontal]
  /// if they need to match a horizontally-scrolling `child`.
  ///
  /// 🔛 [Curtains.textDirection] can manually trigger RTL
  /// (but Curtains does check).
  ///
  /// 🕴 Simple constructor 📜 [Curtains.elevated] uses
  /// [Elevation.asBoxDecoration] to render its decorations;
  /// but feel free to use these static methods, too.
  /// - with 👥 [`package:shadows`](https://pub.dev/packages/shadows)
  /// {@endtemplate}
  const InstantCurtainsDemo({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    // 🕴
    final _shadow = (Elevation.asBoxShadows(24, color: Colors.red) +
            const [
              BoxShadow(color: Colors.yellow, spreadRadius: 10, blurRadius: 10),
              BoxShadow(color: Colors.green, spreadRadius: 5, blurRadius: 25),
            ])
        // Material elevation `List<BoxShadow>`s have 3x shadows
        // - Kept at Material-standard opacities
        //   with `kElevationShadowOpacityRamp`
        // We added 2x shadows which we will now modify.
        //
        // (Simply demonstrating `rampOpacity` util from
        // 👥 Shadows - https://pub.dev/packages/shadows)
        .rampOpacity(kElevationShadowOpacityRamp + [0.2, 0.25]);

    return Expanded(
      child: Curtains.instant(
        scrollDirection: Axis.horizontal, // ↔
        // directionality: TextDirection.rtl, // Manually trigger RTL 🔛
        startCurtain: BoxDecoration(boxShadow: _shadow),
        endCurtain: Elevation.asBoxDecoration(12, color: Colors.purple), // 🕴
        child: ListView(
          scrollDirection: Axis.horizontal, // ↔
          itemExtent: 100.0,
          physics: const BouncingScrollPhysics(),
          children: generatedListHorizontal,
        ),
      ),
    );
  }
}

/// {@macro vertical1}
class CurtainsDemoVertical1 extends StatelessWidget {
  /// {@template vertical1}
  /// ⏰ [Curtains] has intrinsic animation support.
  /// Provide `duration` and/or `curve` or let 📜 [Curtains] default.
  /// {@endtemplate}
  const CurtainsDemoVertical1({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Expanded(
      child: Curtains(
        startCurtain: Elevation.asBoxDecoration(12), // 🕴
        endCurtain: BoxDecoration(boxShadow: Elevation.asBoxShadows(24)), // 🕴
        duration: DURATION, // ⏰
        curve: CURVE, // ⏰
        child: ListView(
          physics: const BouncingScrollPhysics(),
          children: generatedListVertical,
        ),
      ),
    );
  }
}

/// {@macro vertical2}
class CurtainsDemoVertical2 extends StatelessWidget {
  /// {@template vertical2}
  /// [Curtains] and [Curtains.instant] support
  /// full-fat [BoxDecoration]s in lieu of `elevation`.
  /// {@endtemplate}
  const CurtainsDemoVertical2({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Expanded(
      child: Curtains(
        startCurtain: startCurtain, // 🌈
        endCurtain: buildEndCurtain(), // 🌈
        duration: DURATION,
        curve: CURVE,
        child: ListView(
          physics: const BouncingScrollPhysics(),
          children: generatedListVertical,
        ),
      ),
    );
  }
}

/// {@macro horizontal1}
class CurtainsDemoHorizontal1 extends StatelessWidget {
  /// {@template horizontal1}
  /// ```
  /// /// `buildEndCurtain` has a `gradient` that is not visible
  /// /// without [Curtains.spread].
  /// endCurtain: buildFancyEnd(Axis.horizontal), // (explains [Axis] pass)
  /// spread: 25, // Provide "girth" to [_Curtain]s for [Gradient] support.
  ///
  /// /// `BoxDecoration.gradient` clips itself but [BoxShadows], which the
  /// /// simple 📜 [Curtains.elevated] relies on for [Curtains.elevation],
  /// /// need a [ClipRect] in order to not overflow visually.
  /// /// - Manually disable clipping with [clipBehavior] set `Clip.none`
  /// clipBehavior: Clip.none,
  /// ```
  /// {@endtemplate}
  const CurtainsDemoHorizontal1({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Expanded(
      child: Curtains(
        scrollDirection: Axis.horizontal,
        startCurtain: startCurtain,

        /// `buildEndCurtain` has a `gradient` that is not visible
        /// without [Curtains.spread].
        endCurtain: buildEndCurtain(Axis.horizontal), // (explains [Axis] pass)
        spread: 25, // Provide "girth" to [_Curtain]s for [Gradient] support.
        duration: DURATION,
        curve: CURVE,

        /// `BoxDecoration.gradient` clips itself but [BoxShadows], which the
        /// simple 📜 [Curtains.elevated] relies on for [Curtains.elevation],
        /// need a [ClipRect] in order to not overflow visually.
        /// - Manually disable clipping with [clipBehavior] set `Clip.none`
        clipBehavior: Clip.none,
        child: ListView(
          scrollDirection: Axis.horizontal,
          itemExtent: 150.0,
          physics: const BouncingScrollPhysics(),
          children: generatedListHorizontal,
        ),
      ),
    );
  }
}

/// {@macro vertical3}
class CurtainsDemoVertical3 extends StatelessWidget {
  /// {@template vertical3}
  /// ⚖ With a `Sensitivity` [Curtains.sensitivity],
  /// the 📜 [Curtains] scrims will appear later and disappear sooner.
  /// {@endtemplate}
  const CurtainsDemoVertical3({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Expanded(
      child: Curtains(
        startCurtain: startCurtain,
        endCurtain: buildEndCurtain(),
        // `start` appears once scrolled `350` px beyond start
        // `end` appears once scrolled `175` px beyond end
        sensitivity: const Sensitivity(350.0, 175.0),
        duration: DURATION,
        curve: CURVE,
        child: ListView(
          physics: const BouncingScrollPhysics(),
          children: generatedListVertical,
        ),
      ),
    );
  }
}

/// {@macro horizontal2}
class CurtainsDemoHorizontal2 extends StatelessWidget {
  /// {@template horizontal2}
  /// The `startCurtain` and `endCurtain` fields accept more than just
  /// [BoxDecoration]s. Any old [Decoration] will do!
  /// {@endtemplate}
  const CurtainsDemoHorizontal2({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    Decoration buildCurtain({isStart = true}) => ShapeDecoration(
          shadows: [
            BoxShadow(
              color: const Color(0xAA000000),
              spreadRadius: -15,
              blurRadius: 100,
              offset: Offset(isStart ? -75 : 75, 0),
            )
          ],
          image: DecorationImage(
            image: isStart
                ? const NetworkImage('https://i.imgur.com/kvtuWIe.png')
                : const NetworkImage('https://i.imgur.com/hYyYJ0I.png'),
            alignment: Alignment.topCenter,
          ),

          /// Demonstrative
          shape: const RoundedRectangleBorder(
            borderRadius: BorderRadius.all(Radius.elliptical(0, 500)),
          ),
        );

    return Expanded(
      child: Curtains(
        scrollDirection: Axis.horizontal,
        startCurtain: buildCurtain(),
        endCurtain: buildCurtain(isStart: false),
        // spread: 167, // real horizontal resolution
        spread: 100,
        duration: DURATION,
        curve: CURVE,
        child: ListView(
          scrollDirection: Axis.horizontal,
          physics: const BouncingScrollPhysics(),
          children: generatedListHorizontal,
        ),
      ),
    );
  }
}

///
Container buildLeading(Axis axis, int i) => Container(
      decoration: BoxDecoration(
          color: Colors.primaries[Random().nextInt(Colors.primaries.length)],
          border: Border.all(color: Colors.black, width: 2.0)),
      width: (axis == Axis.vertical) ? 50.0 : 75.0,
      height: (axis == Axis.vertical) ? 50.0 : 75.0,
      child: Text(
        '${i + 1}',
        style: TextStyle(
          fontSize: (axis == Axis.vertical) ? 20 : 34,
          fontWeight: FontWeight.bold,
          color: Colors.white,
          height: 1,
        ),
      ),
    );

///
BoxDecoration buildEndCurtain([Axis? axis]) => BoxDecoration(
      // TODO: fix [backgroundBlendMode]
      // backgroundBlendMode: BlendMode.colorDodge,
      gradient: LinearGradient(
        colors: [Colors.green[400]!, Colors.green[900]!.withOpacity(0)],
        begin: (axis == Axis.vertical)
            ? Alignment.bottomCenter
            : Alignment.centerRight,
        end: (axis == Axis.vertical)
            ? Alignment.topCenter
            : Alignment.centerLeft,
      ),
      boxShadow: const [
        BoxShadow(
          color: Colors.blue,
          spreadRadius: 15.0,
          blurRadius: 20.0,
          offset: Offset(0, 15),
        ),
      ],
    );

///
const startCurtain = BoxDecoration(
  boxShadow: [
    BoxShadow(
      color: Color(0xFFFF0000),
      spreadRadius: 1.0,
      blurRadius: 5.0,
    ),
    BoxShadow(
      color: Color(0xBBFF0000),
      spreadRadius: 10.0,
      blurRadius: 30.0,
    ),
    BoxShadow(
      color: Color(0x66FF0000),
      spreadRadius: 25.0,
      blurRadius: 150.0,
    ),
  ],
);

///
final List<Widget> generatedListVertical = List.generate(
  35,
  (i) => Padding(
    padding: const EdgeInsets.only(top: 5.0),
    child: ListTile(
      tileColor: Colors.red.withOpacity(0.05),
      leading: buildLeading(Axis.vertical, i),
      title: titleVertical,
      subtitle: subtitleVertical,
    ),
  ),
);

///
final List<Widget> generatedListHorizontal = List.generate(
  35,
  (i) => Padding(
    padding: const EdgeInsets.only(right: 10.0),
    child: Container(
      color: Colors.red.withOpacity(0.05),
      height: double.infinity,
      child: Column(
        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
        children: <Widget>[
          buildLeading(Axis.horizontal, i),
          titleHorizontal,
          subtitleHorizontal,
        ],
      ),
    ),
  ),
);

///
const titleVertical = Text('Foo Bar Boo Baz', style: TextStyle(fontSize: 20));

///
const subtitleVertical = Text('Lorem ipsum dolor sit amet');

///
const titleHorizontal = Text(
  'Foo\nBar\nBoo\nBaz',
  textAlign: TextAlign.center,
  style: TextStyle(fontSize: 36),
);

///
const subtitleHorizontal = Text(
  '\nLorem\nipsum\ndolor\nsit\namet\nlorem\nipsum\ndolor\nsit\namet',
  textAlign: TextAlign.center,
  style: TextStyle(fontSize: 22),
);

///
const style = TextStyle(fontSize: 20, fontWeight: FontWeight.bold);

///
const headerVertical = SizedBox(
  height: 90,
  child: DecoratedBox(
    decoration: BoxDecoration(color: Color(0x44FF0000)),
    child: Align(
      alignment: Alignment.center,
      child: Text('Vertical ListView 👤 Header', style: style),
    ),
  ),
);

///
const footerVertical = SizedBox(
  height: 60,
  child: DecoratedBox(
    decoration: BoxDecoration(color: Color(0x4400C3FF)),
    child: Align(
      alignment: Alignment.center,
      child: Text('Vertical ListView 🦶 Footer', style: style),
    ),
  ),
);

///
const headerHorizontal = SizedBox(
  width: 80,
  child: DecoratedBox(
    decoration: BoxDecoration(color: Color(0x44FF0000)),
    child: Center(
      child: Text(
        'H\nO\nR\nI\nZ\nO\nN\nT\nA\nL\n\n'
        '👤\n\nH\nE\nA\nD\nE\nR',
        textAlign: TextAlign.center,
        style: style,
      ),
    ),
  ),
);

///
const footerHorizontal = SizedBox(
  width: 60,
  child: DecoratedBox(
    decoration: BoxDecoration(color: Color(0x4400C3FF)),
    child: Center(
      child: Text(
        'H\nO\nR\nI\nZ\nO\nN\nT\nA\nL\n\n'
        '🦶\n\nF\nO\nO\nT\nE\nR',
        textAlign: TextAlign.center,
        style: style,
      ),
    ),
  ),
);
  ''',
    );
  }
}
